v2/blockstore/readwrite.go

```diff
@@ -59,50 +66,164 @@ func WithIndexPadding(p uint64) Option {
 	b.dedupCids = true
 }
 
-// NewReadWrite creates a new ReadWrite at the given path with a provided set of root CIDs as the car roots.
+// WithResumption sets whether the blockstore should resume on a file produced by a ReadWrite
+// blockstore on which ReadWrite.Finalize was not called.
+//
+// This option is set to false by default, i.e. disabled. When disabled, the file at path must not
+// exist, otherwise an error is returned upon ReadWrite blockstore construction.
+//
+// When this option is set to true the existing data frames in file are re-indexed, allowing the
+// caller to continue putting any remaining blocks without having to re-ingest blocks for which
+// previous ReadWrite.Put returned successfully.
+//
+// Resumption is only allowed on files that satisfy the following criteria:
+//   1. start with a complete CAR v2 car.Pragma.
+//   2. contain a complete CAR v1 data header with identical roots, starting at offset specified by
+//      WithCarV1Padding, followed by zero or more complete data frames. If any corrupt data frames
+//      are present the resumption will fail. Note, it is important that new instantiations of
+//      ReadWrite blockstore with resumption enabled use the same WithCarV1Padding option, since
+//      this option is used to locate the offset at which the data payload starts.
+//   3. have not been produced by a ReadWrite blockstore that was finalized, i.e. call to
+//      ReadWrite.Finalize returned successfully.
+func WithResumption(enabled bool) Option {
+	return func(b *ReadWrite) {
+		b.resume = enabled
+	}
+}
+
+// NewReadWrite creates a new ReadWrite at the given path with a provided set of root CIDs and options.
 func NewReadWrite(path string, roots []cid.Cid, opts ...Option) (*ReadWrite, error) {
-	// TODO support resumption if the path provided contains partially written blocks in v2 format.
 	// TODO either lock the file or open exclusively; can we do somethign to reduce edge cases.
-	f, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0o666)
-	if err != nil {
-		return nil, fmt.Errorf("could not open read/write file: %w", err)
+
+	b := &ReadWrite{
+		header: carv2.NewHeader(0),
 	}
 
 	indexcls, ok := index.BuildersByCodec[index.IndexInsertion]
 	if !ok {
 		return nil, fmt.Errorf("unknownindex  codec: %#v", index.IndexInsertion)
 	}
-	idx := (indexcls()).(*index.InsertionIndex)
+	b.idx = (indexcls()).(*index.InsertionIndex)
 
-	b := &ReadWrite{
-		f:      f,
-		idx:    idx,
-		header: carv2.NewHeader(0),
-	}
 	for _, opt := range opts {
 		opt(b)
 	}
-	b.carV1Writer = internalio.NewOffsetWriter(f, int64(b.header.CarV1Offset))
-	carV1Reader := internalio.NewOffsetReadSeeker(f, int64(b.header.CarV1Offset))
-	b.ReadOnly = ReadOnly{backing: carV1Reader, idx: idx}
-	if _, err := f.WriteAt(carv2.Pragma, 0); err != nil {
-		return nil, err
-	}
 
-	v1Header := &carv1.CarHeader{
-		Roots:   roots,
-		Version: 1,
+	fFlag := os.O_RDWR | os.O_CREATE
+	if !b.resume {
+		fFlag = fFlag | os.O_EXCL
```

Comment:
don't we want to be exclusive in the resume case as well?

Reply:
No, because I explicitly want `ReadWrite` constructor to return an error, when resumption is set to false but the file already exists. I have documented this on resumption option [here](https://github.com/ipld/go-car/blob/debbc7edddb6fdf161663d7e6f7311df9fe763f9/v2/blockstore/readwrite.go#L76-L77). 

The rationale is to avoid overriding existing files by accident.

Reply:
Following the same rationale, altered the code slight to avoid creating files when resumption is enabled in https://github.com/ipld/go-car/pull/147/commits/e12754dd76e8d076f6302c13ed0127afd2de63b1.